Implementation of Multiprobe LSH

mentekid opened the pull request:
I have also included the deterministic test changes so I can test my code better. You should probably ignore those.

There's a weird bug, I think it's where I create the hashMat but I'm not sure.

mentekid commented:
I think I fixed it, take a look.
Next stop adding tests

rcurtin commented on src/mlpack/methods/lsh/lsh_search.hpp:
```diff
@@ -143,7 +159,8 @@ class LSHSearch
               arma::mat& distances,
-              const size_t numTablesToSearch = 0);
+              const size_t numTablesToSearch = 0,
+              size_t T = 0);
```
Can you document what the `T` parameter is here?  We should probably mark it const too if we can.

mentekid replied:
You're right, I forgot to add that, I was only documenting T in ReturnIndicesFromTables.

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -362,27 +652,65 @@ void LSHSearch<SortPolicy>::ReturnIndicesFromTable(
   Log::Assert(hashVec.n_elem == numTablesToSearch);
 
+  // Compute hashVectors of additional probing bins
+  arma::mat hashMat;
```
Two thoughts:

 * couldn't this be an `arma::Mat<size_t>`?  I'm not sure why the original `hashVec` didn't hold a `size_t` internally.
 * we could probably save a copy and speed things up a little if we never use the original `hashVec` and instead use `hashMat` and set its size to `(T + 1, numTablesToSearch)`.

mentekid replied:
**On your first point:** I wondered too when I first saw this - I thought it had something to do with operator precedence but [typecasting precedes modding](http://en.cppreference.com/w/c/language/operator_precedence), so I'm fixing that.

**On your second point:** In the beginning I started doing it like that to avoid the copy, but the armadillo spans confused me so I went around it. I'll try to find a way to do it the way you say.

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -362,27 +652,65 @@ void LSHSearch<SortPolicy>::ReturnIndicesFromTable(
+
+      // map the probing bin to second hash table bins
+      hashMat.col(i) = additionalProbingBins.t() * secondHashWeights;
+      //hashMat.col(i) = dummyBins;
```
You can probably remove the commented out debugging code :)

mentekid replied:
yup! done

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -339,12 +340,301 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+  std::pair<double, size_t> pair0( perturbationScore(Ao, scores), 0 );
+  minHeap.push(pair0);
+
+  double prevScore = 0; // store score of smallest inserted vector (for assert)
```
If this variable is here just for an assert, then I think once the tests show the algorithm is working then we can remove this.

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -339,12 +340,301 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+        std::pair<double, size_t> shifted(
+            perturbationScore(As, scores),
+            perturbationSets.size() - 1); // (score, position) pair for shift
+        minHeap.push(shifted);
```
Consider also using `make_pair`:

```
minHeap.push(std::make_pair(perturbationScore(As, scores), perturbationSets.size() - 1));
```

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -339,12 +340,301 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+        std::pair<double, size_t> p2){
+      //only compare the double values
+      return p1.first > p2.first;
+    }
```
Are you sure `std::greater` will not work here instead of this class?  By default for `std::pair` that will take into account the second element, but I think that in your case ordering based on the second element doesn't really matter.

mentekid replied:
You are right - I started implementing that, but then somewhere in the middle found a bug with my initial code so I never got around to throwing CompareGreater away.

Yes, in the very unlikely event where two scores are exactly the same, it doesn't really matter which one gets priority anyway - so ordering the second element is as arbitrary as anything else.

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -339,12 +340,301 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+      max = A[i];
+      max_pos = i;
+    }
+  }
```
It's unfortunate that `std::vector` has to be used for `A`; if we could use Armadillo objects there then this function could be way simpler.  But after looking through the code I am not sure there is a way that we can use an Armadillo object instead of `std::vector`, since all the `A` vectors will be of different (and changing) sizes.  Correct me if I am wrong about that. :)

mentekid replied:
What if instead of A storing indices to some of the 2*numProj elements, each A was 2*numProj long and stored 1 if the corresponding element is "included" and 0 otherwise?
So instead of A = [1, 5] we would have A = [0 1 0 0 0 1 0 ... 0]
If arma::Row<bool> compresses its elements then this will avoid both the extra space and the use of std::vector.

mentekid replied:
So, I was wrong in that we cannot create an arma::Row<bool>. Armadillo's smallest type is a char, which compresses more than ints but still uses 8 times as much space as we would actually need.

If we're willing to complicate the implementation a little, we could use [std::bitset](http://www.cplusplus.com/reference/bitset/bitset/).

rcurtin replied:
I'd just pick `std::vector<bool>`, since that will pack things the same way `bitset` would, but does not have the same compile-time length requirement.

rcurtin replied:
Ah, sorry, I didn't see a later comment of yours.  If the `std::vector<bool>` solution isn't to your liking and it doesn't give any performance change, then it's up to you which implementation you want to go with.  Personally I'd probably pick the one that gives clearer code, but which one that is is up to you.

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -339,12 +340,301 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+{
+  // Stack allocation and initialization to 0 (bool check[numProj] = {0}) made
+  // some compilers complain. We use new just to be safe.
+  bool *check = new bool[numProj]();
```
Maybe use `std::vector<bool>` here?  It avoids the need for manual memory management and I think that most compilers/stdlib implementations will pack the booleans so the required amount of memory will be smaller.  I can't remember the exact details at the moment.

mentekid replied:
Done. I think what you said makes more sense.

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -339,12 +340,301 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+  }
+
+  // sort everything in increasing order
+  arma::Col<long long unsigned int> sortidx = arma::sort_index(scores);
```
Why `long long unsigned int`?  Either `size_t` or possibly `arma::uword` (I prefer the former but sometimes it is not possible) should work here.

mentekid replied:
Yep -  I had trouble when using size_t, the compiler didn't like it. I replaced it with arma::uword. I'm not sure why size_t is not allowed...

rcurtin commented:
Ok, I made all the comments I have.  I think it is ready to merge once we address those comments.

My only other note is, could you go through and fix the style?  If not I can do it when I merge, it just saves me a few minutes. :)

mentekid commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -362,27 +652,65 @@ void LSHSearch<SortPolicy>::ReturnIndicesFromTable(
+  arma::mat hashMat;
+  if (T > 0)
+  {
+    hashMat.zeros(T, numTablesToSearch);
```
If I concatenate a (T x numTablesToSearch) matrix with a (1 x numTablesToSearch) vector, won't I get a T+1 row matrix? 

I understood from armadillo documnetation that `join_vert` resizes hashMat. Am I throwing the last vector away like that?

mentekid commented:
I merged the changes from previous commits so travis/appveyor can run the tests.

I found a weird behavior, so please check if you can find any bugs in my code:
When running the "cheap run" test (expecting recall to be very low) with the previous parameters, specifically numProj=1000, made the test to fail. I set numProj=100 and now the test passes. I'm not sure if this is an actual bug or simply caused by the dataset - it never failed before.

If I broke something, let me know :)

mentekid commented:
So, I fixed some bugs and implemented all your comments. I'm not sure if the `std::vector<bool>` code is simpler (or even faster) than the initial code. Some tests I ran gave pretty similar results, but that may be because of the parameters.

I tried to fix as many style issues as I could find, let me know what I didn't find because I think if you fix it I won't remember it for next time :)

zoq commented on src/mlpack/methods/lsh/lsh_search.hpp:
```diff
@@ -320,6 +345,25 @@ class LSHSearch
+                            const arma::vec &queryCodeNotFloored,
+                            const size_t T,
+                            arma::mat &additionalProbingBins) const;
+
```
It's a matter of preference, but to be consistent with the rest of the code: when declaring a variable, place the `*` or `&` adjacent to the type.

mentekid replied:
thank you, I didn't notice that.

mentekid commented:
So, there is an error in my logic behind MultiprobeDeterministicTest. I think it's an error in the test and not in the code. Maybe it would be better to remove that test completely. Here's where I went wrong.

When creating the four clusters of points, what I had in mind was this:
![PointCluster.png](http://i.imgur.com/JcozL2s.png)

So the groups of blue points would be assigned the same code. This doesn't happen because of the random offsets added before flooring, so the actual image is more like:
![PointCluster_offset.png](http://i.imgur.com/VJggrVB.png)

meaning, each cluster is not confined to only one code as I initially thought. This makes this test fail, because for example the red (query) point will have neighbors even without multiprobe but will obviously have more neighbors when multiprobe is used. This is something we actually check in the randomized check, so I removed this test until I could find a better one.

Note that the single-probe deterministic test still works - the clusters don't mix

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -331,12 +333,424 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+// it contains two (or more) actions for the same dimension or dimensions that
+// are larger than the queryCode's dimensions.
+inline bool perturbationValid(const std::vector<bool>& A,
+                              const size_t numProj)
```
Another small comment---we should name this function `PerturbationValid` (and the others should have a capital first letter too).  It might make more sense to put them inside of the `LSH` class just for the sake of organization, but whatever you want to do there is fine.

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -331,12 +333,424 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+  if ( maxPos + 1 < A.size()) // otherwise, this is an invalid vector 
+  {
+    A[maxPos] = 0;
+    A[maxPos+1] = 1;
```
I'm going to be picky about style again, this should be `maxPos + 1`, and further up it should be `if (maxPos + 1` and above that, `maxPos = i`. :)

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -331,12 +333,424 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+      additionalProbingBins(positions(Ai[i]), pvec) += actions(Ai[i]);
+
+  }
+  */
```
If you like the new implementation, we can go ahead and remove the old one. :)  There are some other commented-out implementations above that we can remove too.

rcurtin commented on src/mlpack/tests/lsh_test.cpp:
```diff
@@ -466,6 +466,143 @@ BOOST_AUTO_TEST_CASE(DeterministicNoMerge)
+          numProbes[p]);
+
+      // compute recall
+      double recall = LSHSearch<>::ComputeRecall(lshNeighbors, groundTruth); //TODO: change to LSHSearch::ComputeRecall??
```
I think the TODO is no longer accurate. :)

rcurtin commented:
I think that the offsets have a maximum, right?  So every point from, e.g., the lower left cluster will fall not in the lower-leftmost bin, but instead in (I think) the 4 (or 9?) bins closest to the center of the lower-leftmost bin.  (Or something like that.)  If that's the case, we could then put the query point outside of the furthest possible bin, and then set a higher number of probes to ensure that at least one of the 4 (or 9) bins are checked.

My thought there isn't complete, but I hope it is coherent enough to make sense.  Let me know if I can explain further.

The rest of the code itself looks good, so when you decide what to do with `std::vector<bool>` or the previous implementation and figure out this test, I think it's good to go---unless there is something else I overlooked.

I'll go through and point out the style, but it will be a lot of comments, so remember, you asked for it. :)

rcurtin commented on src/mlpack/methods/lsh/lsh_main.cpp:
```diff
@@ -64,6 +64,8 @@ PARAM_INT("tables", "The number of hash tables to be used.", "L", 30);
 PARAM_DOUBLE("hash_width", "The hash width for the first-level hashing in the "
     "LSH preprocessing. By default, the LSH class automatically estimates a "
     "hash width for its use.", "H", 0.0);
+PARAM_INT("num_probes", "Number of additional probes for Multiprobe LSH"
```
We should probably add a semicolon after "LSH" and decapitalize the following "If" so it is one complete sentence. Or something like that.

rcurtin commented on src/mlpack/methods/lsh/lsh_main.cpp:
```diff
@@ -137,6 +139,7 @@ int main(int argc, char *argv[])
   const size_t numProj = CLI::GetParam<int>("projections");
   const size_t numTables = CLI::GetParam<int>("tables");
   const double hashWidth = CLI::GetParam<double>("hash_width");
+  const size_t numProbes = CLI::GetParam<int>("num_probes");
```
Do you need to do an explicit cast here to avoid a warning?  I am not sure, I think it depends on the compiler, but usually I will do something like `(size_t) CLI::GetParam<int>("num_probes");`.

rcurtin commented on src/mlpack/methods/lsh/lsh_search.hpp:
```diff
@@ -19,7 +19,23 @@
+ * }
+ *
  */
+
```
This line seems to be extra and can be removed.

rcurtin commented on src/mlpack/methods/lsh/lsh_search.hpp:
```diff
@@ -259,11 +279,16 @@ class LSHSearch
    *    hashing the query into all the hash tables and eventually into
    *    multiple buckets of the second hash table.
+   * @param numTablesToSearch The number of tables to perform the search in. If
+   * 0, all tables are searched.
```
Can you space these second lines four spaces in like the other lines?

rcurtin commented on src/mlpack/methods/lsh/lsh_search.hpp:
```diff
@@ -320,6 +345,25 @@ class LSHSearch
+   * query.
+   * @param T number of additional probing bins.
+   * @param additionalProbingBins matrix. Each column will hold one additional
+   * bin.
```
The same applies here about the line wrap spacing.

rcurtin commented on src/mlpack/methods/lsh/lsh_search.hpp:
```diff
@@ -320,6 +345,25 @@ class LSHSearch
+  void GetAdditionalProbingBins(const arma::vec& queryCode,
+                            const arma::vec& queryCodeNotFloored,
+                            const size_t T,
+                            arma::mat& additionalProbingBins) const;
```
The spacing seems off here, the second/third/fourth arguments should align with the first.

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -9,6 +9,7 @@
 
 #include <mlpack/core.hpp>
 
+
```
This seems like an extra line too.

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -331,12 +333,423 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+    if (A[i]) // marked true
+      maxPos = i;
+
+  if ( maxPos + 1 < A.size()) // otherwise, this is an invalid vector
```
`if (maxPos + 1 < A.size())`

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -331,12 +333,423 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+      continue;
+
+    // If dimesnion is unseen thus far, mark it as seen.
+    if ( check[i % numProj] == false )
```
Extra spaces on the inside of the parentheses; we should go with `if (check[i % numProj] == false)`.

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -331,12 +333,423 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+// Compute additional probing bins for a query
+template<typename SortPolicy>
+void LSHSearch<SortPolicy>::GetAdditionalProbingBins(
+                            const arma::vec& queryCode,
```
The spacing seems odd here; if we can't fit the arguments on the same line as the function name, then we should just indent them four spaces from the start of the line.

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -331,12 +333,423 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+                            const size_t T,
+                            arma::mat& additionalProbingBins) const
+{
+
```
Extra line :)

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -331,12 +333,423 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+
+
+  // Acting dimension vector shows which coordinate to transform according to
+  // actions described by actions vector
```
I always feel so pedantic writing this, but can you finish the sentence with a period?  I think it's important to have complete sentences in documentation for clarity, so that people for whom English isn't the first language can have a better understanding.

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -331,12 +333,423 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+  // Special case: No need to create heap for 1 or 2 codes.
+  if (T <= 2)
+  {
+
```
Extra line

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -331,12 +333,423 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+    size_t minloc2 = 0;
+    for (size_t s = 0; s < (2 * numProj); ++s) // here we can't start from 1
+    {
+      if ( minscore2 > scores[s] && s != minloc) //second smallest
```
Extra space after opening parenthesis.

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -331,12 +333,423 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+  // General case: more than 2 perturbation vectors require use of minheap.
+
+  // Sort everything in increasing order.
+  arma::Col<long long unsigned int> sortidx = arma::sort_index(scores);
```
I think this was mentioned somewhere else, so forgive any redundancy, but probably `arma::uword` should be used here instead of `long long unsigned int`.  You can just use `arma::uvec` since that is a typedef of `arma::Col<arma::uword>`.

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -331,12 +333,423 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+
+      // Shift operation on Ai (replace max with max+1).
+      std::vector<bool> As = Ai;
+      if ( perturbationShift(As) && perturbationValid(As, numProj) )
```
Extra space after opening parenthesis and before closing parenthesis.

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -331,12 +333,423 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+        perturbationSets.push_back(As); // add shifted set to sets
+        minHeap.push( 
+            std::make_pair(
+              perturbationScore(As, scores), 
```
It's fine to indent this after the `make_pair` is already indented, but I would go an additional four spaces in if you are going to do that.  Personally, here, I would just do this:

```
minHeap.push(std::make_pair(perturbationScore(As, scores),
                                                perturbationSets.size() - 1));
```

(I doubt that will line up right, but the idea is to line up `perturbationScore(As, scores)` with `perturbationSets.size() - 1`, and, no need to have an extra line to close the whole thing.

mentekid replied:
I always get confused when I have too many nested functions - that's why I space it out. I changed this to
```
minHeap.push(
    std::make_pair(PerturbationScore(As, scores),
    perturbationSets.size() - 1)
    );
```

I think this way it's clear what the two parts of the pair are, but maybe that's just me. I could change it to what you suggested.

rcurtin replied:
Sure, that's fine, but do you think it is not too unclear if you include the final `);` on the third line?  I think there is a lot in the style guide that is maybe not clear, so I am happy to go either way if we can eventually codify what is okay and what is not, so we can update the style guide. :)

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -331,12 +333,423 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+            std::make_pair(
+              perturbationScore(Ae, scores),
+              perturbationSets.size() - 1)
+            );
```
Same comment as above applies here about the spacing.

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -331,12 +333,423 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+
+      // Expand operation on Ai (add max+1 to set).
+      std::vector<bool> Ae = Ai;
+      if ( perturbationExpand(Ae) && perturbationValid(Ae, numProj) ) 
```
Extra space after the opening parenthesis and before the closing parenthesis.

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -331,12 +333,423 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+            );
+      }
+
+    }while (! perturbationValid(Ai, numProj)  );//Discard invalid perturbations
```
This is weird spacing; should be `} while (!PerturbationValid(Ai, numProj)); // Discard invalid perturbations.`

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -331,12 +333,423 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+    for (size_t pos = 0; pos < Ai.size(); ++pos)
+      // If Ai[pos] is marked, add action to probing vector.
+      additionalProbingBins(positions(pos), pvec) 
+          += Ai[pos] ? actions(pos) : 0;   }
```
I'm confused here, I am not sure what this closing brace is for.  Maybe it should be on the next line?

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -331,12 +333,423 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+          += Ai[pos] ? actions(pos) : 0;   }
+}
+
+
```
Extra line, we only need one between methods.

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -354,29 +767,61 @@ void LSHSearch<SortPolicy>::ReturnIndicesFromTable(
-  arma::rowvec hashVec = secondHashWeights.t() * arma::floor(allProjInTables);
+    queryCodesNotFloored.unsafe_col(i) = projections.slice(i).t() * queryPoint;
+  queryCodesNotFloored += offsets.cols(0, numTablesToSearch - 1);
+  allProjInTables = arma::floor(queryCodesNotFloored/hashWidth);
```
We should put a space on either side of the division operator `/`.

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -354,29 +767,61 @@ void LSHSearch<SortPolicy>::ReturnIndicesFromTable(
+  queryCodesNotFloored += offsets.cols(0, numTablesToSearch - 1);
+  allProjInTables = arma::floor(queryCodesNotFloored/hashWidth);
+
+
```
Extra line

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -354,29 +767,61 @@ void LSHSearch<SortPolicy>::ReturnIndicesFromTable(
+  // the secondHashTable using the secondHashWeights.
+  hashMat.row(0) = 
+    arma::conv_to< arma::Row<size_t> >:: // floor by typecasting to size_t
+    from( secondHashWeights.t() * allProjInTables );
```
I think this could fit a bit more concisely on one (or probably two) lines; there's no need for the spaces between the `<>` and there are extra spaces after `from(` and before the closing `);`.

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -354,29 +767,61 @@ void LSHSearch<SortPolicy>::ReturnIndicesFromTable(
+  hashMat.row(0) = 
+    arma::conv_to< arma::Row<size_t> >:: // floor by typecasting to size_t
+    from( secondHashWeights.t() * allProjInTables );
+  // mod to compute 2nd-level codes
```
Again I feel really pedantic saying this but if this comment can be a complete sentence it would be more clear. :)

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -354,29 +767,61 @@ void LSHSearch<SortPolicy>::ReturnIndicesFromTable(
 
-  for (size_t i = 0; i < hashVec.n_elem; i++)
-    hashVec[i] = (double) ((size_t) hashVec[i] % secondHashSize);
+  // Compute hash codes of additional probing bins
```
Missing period :)

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -354,29 +767,61 @@ void LSHSearch<SortPolicy>::ReturnIndicesFromTable(
+      // the primary hash table).
+      hashMat(arma::span(1, T), i) = // Compute code of rows 1:end of column i
+        arma::conv_to< arma::Col<size_t> >:: // floor by typecasting to size_t
+        from( secondHashWeights.t() * additionalProbingBins );
```
The same comments apply from earlier; I think this can be shortened and extra spaces removed.

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -398,15 +843,22 @@ void LSHSearch<SortPolicy>::ReturnIndicesFromTable(
+      for (size_t p = 0; p < T + 1; ++p) // for entire probing sequence
+      {
+
+        // get the sequence code
```
Super-pedant mode--- capitalization and period. :)

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -477,6 +933,23 @@ void LSHSearch<SortPolicy>::Search(const arma::mat& querySet,
+    Teffective = ( 1 << numProj ) - 1;
+    Log::Warn << "Requested " << T << 
+      " additional bins are more than theoretical maximum. Using " <<
+      Teffective << " instead." << std::endl;
```
I like to make these lines as long as I can by splitting the strings up, and starting new lines with `<<`.  Whatever you want to do is fine, but the new lines should at least be indented four characters instead of two.

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -477,6 +933,23 @@ void LSHSearch<SortPolicy>::Search(const arma::mat& querySet,
+  size_t Teffective = T;
+  if (T > ( (size_t) ( (1 << numProj) - 1) ) )
+  {
+    Teffective = ( 1 << numProj ) - 1;
```
There are a bunch of extra spaces in these three lines:

```
if (T > ((size_t) ((1 << numProj) - 1)))
{
  Teffective = (1 << numProj) - 1;
```

Technically variables should have names that start with lowercase letters, so I would have gone with `t` and `effectiveT` instead.

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -477,6 +933,23 @@ void LSHSearch<SortPolicy>::Search(const arma::mat& querySet,
+    Log::Info << "Running multiprobe LSH with " << Teffective <<
+      " additional probing bins per table per query."<< std::endl;
+
+
```
Extra line

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -477,6 +933,23 @@ void LSHSearch<SortPolicy>::Search(const arma::mat& querySet,
+  }
+
+  // If the user set multiprobe, log it
+  if (T > 0)
```
Actually should we be checking `Teffective` here?

rcurtin commented on src/mlpack/methods/lsh/lsh_search_impl.hpp:
```diff
@@ -514,14 +988,31 @@ void LSHSearch<SortPolicy>::
+  }
+
+  // If the user set multiprobe, log it
+  if (T > 0)
```
Same here, should we be checking `Teffective`?

rcurtin commented on src/mlpack/tests/lsh_test.cpp:
```diff
@@ -466,6 +466,145 @@ BOOST_AUTO_TEST_CASE(DeterministicNoMerge)
+        numTables,
+        hashWidth,
+        secondHashSize,
+        bucketSize);
```
You could probably fit this on just one or two lines, but it does not matter to me.  Also the comment could be capitalized and have a period after it. :)

rcurtin commented on src/mlpack/tests/lsh_test.cpp:
```diff
@@ -466,6 +466,145 @@ BOOST_AUTO_TEST_CASE(DeterministicNoMerge)
+        bucketSize);
+
+    double prevRecall = 0;
+    // search with varying number of probes
```
Capitalization, period

rcurtin commented on src/mlpack/tests/lsh_test.cpp:
```diff
@@ -466,6 +466,145 @@ BOOST_AUTO_TEST_CASE(DeterministicNoMerge)
+    for (size_t p = 0; p < probeTrials; ++p)
+    {
+      arma::Mat<size_t> lshNeighbors;
+      arma::mat lshDistances; //move outside of loop for speed?
```
No need to move it outside the loop, it should not make any noticeable difference.  No need to keep the comment, in my opinion.

rcurtin commented on src/mlpack/tests/lsh_test.cpp:
```diff
@@ -466,6 +466,145 @@ BOOST_AUTO_TEST_CASE(DeterministicNoMerge)
+          lshNeighbors, 
+          lshDistances, 
+          0, 
+          numProbes[p]);
```
This might all fit on one line.

rcurtin commented on src/mlpack/tests/lsh_test.cpp:
```diff
@@ -466,6 +466,145 @@ BOOST_AUTO_TEST_CASE(DeterministicNoMerge)
+
+        // ... and should ideally increase it a bit
+        if (recall > prevRecall + epsilonIncrease)
+         foundIncrease = true;
```
This is weird, it's only one space in, not two.

rcurtin commented on src/mlpack/tests/lsh_test.cpp:
```diff
@@ -466,6 +466,145 @@ BOOST_AUTO_TEST_CASE(DeterministicNoMerge)
+  // Test that q1 simple search comes up empty
+  lshTest.Search(q1, k, neighbors, distances);
+  cout << neighbors << endl;
+  BOOST_REQUIRE( arma::all(neighbors.col(0) == N) );
```
Extra spaces here... `BOOST_REQUIRE(arma::all(neighbors.col(0) == N));`

rcurtin commented on src/mlpack/tests/lsh_test.cpp:
```diff
@@ -466,6 +466,145 @@ BOOST_AUTO_TEST_CASE(DeterministicNoMerge)
+  // Searching with 3 additional probing bins should find neighbors
+  lshTest.Search(q1, k, neighbors, distances, 0, 3);
+  cout << neighbors << endl;
+  BOOST_REQUIRE( arma::all(neighbors.col(0) == N || neighbors.col(0) < 10) );
```
Extra spaces here too... `BOOST_REQUIRE(arma::all(neighbors.col(0) == N || neighbors.col(0) < 10));`

rcurtin commented on src/mlpack/tests/lsh_test.cpp:
```diff
@@ -466,6 +466,145 @@ BOOST_AUTO_TEST_CASE(DeterministicNoMerge)
+  LSHSearch<> lshTest(rdata, projections,
+                      hashWidth, secondHashSize, bucketSize);
+
+
```
Extra line

rcurtin commented on src/mlpack/tests/lsh_test.cpp:
```diff
@@ -466,6 +466,145 @@ BOOST_AUTO_TEST_CASE(DeterministicNoMerge)
+  arma::Mat<size_t> neighbors;
+  arma::mat distances;
+
+  // Test that q1 simple search comes up empty
```
A lot of the comments in this test don't have capitalization or periods or aren't complete sentences.

rcurtin commented:
Ok, I went through the style... I feel really annoying and pedantic... so sorry for the huge number of emails and notifications. :)  It's all simple stuff that I'm happy to change after merge, but if you'd like to do it now it could save me a few minutes.

mentekid commented:
I went through your comments and fixed them. Don't feel annoying - I think the best way for me to remember these are to implement them a few times. Plus, there's a reason it's easy for people to start contributing - if the code was a mess, nobody would like to dive in.
So, sorry for wasting your time, hopefully it will save you more in the long run :)

mentekid commented:
> I think that the offsets have a maximum, right? So every point from, e.g., the lower left cluster will fall not in the lower-leftmost bin, but instead in (I think) the 4 (or 9?) bins closest to the center of the lower-leftmost bin. (Or something like that.) If that's the case, we could then put the query point outside of the furthest possible bin, and then set a higher number of probes to ensure that at least one of the 4 (or 9) bins are checked.

You're right, the offsets have a maximum and better yet they're actually positive - so the translation they cause can only be to the right and upwards, and because of their maximum size (equal to hash width) they can move points up to one "block".


For the figure, let's take the C2 cluster centered at (3.5, 3.5).
That means we could simply replace the red point with one (say q1) directly under C2, as well as one inside C2's block (say q2). We can access the offsets from the LSHSearch object so we can actually place these wherever they "should be" to get hashed properly.
That way:

 * q1 should find no neighbors without multiprobe, and the neighbors that weren't translated away (so, hashed to C2's original block around) with 1 probe.
* q2 should find some neighbors inside C2 and should find all neighbors with 3 additional probes.

I'll start writing this and see how it goes.

mentekid commented:
So, I noticed something weird which I hadn't noticed before. If you can reproduce it on your system then there might be something wrong either with our implementation or the cheap LSH run test.

I expected that setting L=1 and K=100 (numTables, numProj respectively) will never lead to high recall. This is not the case though, some times this returns the entire dataset and has 100% recall.

For example, run this a few times:
```
bin/mlpack_lsh -r ~/Projects/data/csv/iris_r.csv -q ~/Projects/data/csv/iris_q.csv -t ~/Projects/data/csv/iris_t.csv -K 40 -L 1 -H 0.00001 -k 32 -n garbage.csv -v
```

Does it ever return recall==100? Intuitively, it shouldn't... Have we messed something up or is it just that iris is a weird dataset? I never saw that with sift.

rcurtin commented:
Okay, so the weird bug you mentioned in your previous comment is now fixed (eea2aa4 and e6bc4b4).  If you are happy with the deterministic test now, I can go ahead and merge.  I believe that RecallTest failed in the Travis build due to the casting bug, so I am not worried that it has failed here.

rcurtin commented on src/mlpack/tests/lsh_test.cpp:
```diff
@@ -539,31 +539,28 @@ BOOST_AUTO_TEST_CASE(MultiprobeTest)
-/*
 BOOST_AUTO_TEST_CASE(MultiprobeDeterministicTest)
 {
   math::RandomSeed(std::time(NULL));
```
Ah, we should remove this before merge, so that tests are reproducible for a given compilation.

rcurtin commented on src/mlpack/tests/lsh_test.cpp:
```diff
@@ -539,31 +539,28 @@ BOOST_AUTO_TEST_CASE(MultiprobeTest)
   GetPointset(N, rdata);
 
-  const int k = N/2;
+  const int k = N/4;
```
A couple more style comments, this should be `N / 4` ("horizontal whitespace"). :)

rcurtin commented on src/mlpack/tests/lsh_test.cpp:
```diff
@@ -579,25 +576,43 @@ BOOST_AUTO_TEST_CASE(MultiprobeDeterministicTest)
+  // Construct q2 so it is hashed near the center of C2.
+  arma::mat q2;
+  q2 << 3.6 << arma::endr << 3.6;
+  q2-= offsets;
```
Missing space after `q2` here.

rcurtin commented on src/mlpack/tests/lsh_test.cpp:
```diff
@@ -579,25 +576,43 @@ BOOST_AUTO_TEST_CASE(MultiprobeDeterministicTest)
+      neighbors.col(0) >= 10 && neighbors.col(0) < 20));
 }
-*/
+
```
Extra line

mentekid commented:
I merged eea2aa4 and e6bc4b4 and implemented your style comments.

On the 
```
function(function(
    args
)); 
```
vs
```
function(function(
    args));
```

I ended up going with the 2nd one. I find them equally readable but the second one is shorter.
